File: src/main/java/io/polygloat/dtos/PublicConfigurationDTO.java
package io.polygloat.dtos;

import io.polygloat.configuration.polygloat.PolygloatProperties;

public class PublicConfigurationDTO {
    private boolean authentication;

    private AuthMethodsDTO authMethods;

    private boolean passwordResettable;

    private boolean allowRegistrations;

    public PublicConfigurationDTO(PolygloatProperties configuration) {
        this.authentication = configuration.getAuthentication().getEnabled();
        if (authentication) {
            authMethods = new AuthMethodsDTO(new GithubPublicConfigDTO(configuration.getAuthentication().getGithub().getClientId()));
        }
        passwordResettable = configuration.getAuthentication().getNativeEnabled();
        allowRegistrations = configuration.getAuthentication().getRegistrationsAllowed();
    }

    public boolean isAuthentication() {
        return this.authentication;
    }

    public AuthMethodsDTO getAuthMethods() {
        return this.authMethods;
    }

    public boolean isPasswordResettable() {
        return this.passwordResettable;
    }

    public boolean isAllowRegistrations() {
        return this.allowRegistrations;
    }

    public static class AuthMethodsDTO {
        private GithubPublicConfigDTO github;

        public AuthMethodsDTO(GithubPublicConfigDTO github) {
            this.github = github;
        }

        public GithubPublicConfigDTO getGithub() {
            return this.github;
        }
    }

    public static class GithubPublicConfigDTO {
        private boolean enabled;

        private String clientId;

        public GithubPublicConfigDTO(String clientId) {
            this.clientId = clientId;
            this.enabled = clientId != null && !clientId.isEmpty();
        }

        public boolean isEnabled() {
            return this.enabled;
        }

        public String getClientId() {
            return this.clientId;
        }
    }
}

File: src/main/java/io/polygloat/dtos/PublicConfigurationDTO.kt
package io.polygloat.dtos

import io.polygloat.configuration.polygloat.PolygloatProperties

@Suppress("MemberVisibilityCanBePrivate", "unused")
class PublicConfigurationDTO(properties: PolygloatProperties) {
    val isAuthentication: Boolean = properties.authentication.enabled
    var authMethods: AuthMethodsDTO? = null
    val isPasswordResettable: Boolean
    val isAllowRegistrations: Boolean
    val screenshotsUrl = properties.screenshotsUrl
    val maxUploadFileSize = properties.maxUploadFileSize
    val clientSentryDsn = if (properties.sentry.enabled) properties.sentry.clientDsn else null
    val needsEmailVerification = properties.authentication.needsEmailVerification

    class AuthMethodsDTO(val github: GithubPublicConfigDTO)
    class GithubPublicConfigDTO(val clientId: String?) {
        val isEnabled: Boolean = clientId != null && clientId.isNotEmpty()
    }

    init {
        if (isAuthentication) {
            authMethods = AuthMethodsDTO(GithubPublicConfigDTO(properties.authentication.github.clientId))
        }
        isPasswordResettable = properties.authentication.nativeEnabled
        isAllowRegistrations = properties.authentication.registrationsAllowed
    }
}